scaaml/stats/snr.py

```diff
@@ -0,0 +1,156 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Compute SNR. Inspired by the tutorial
+https://github.com/newaetech/chipwhisperer-jupyter/blob/master/archive/PA_Intro_3-Measuring_SNR_of_Target.ipynb
+"""
+from collections import defaultdict
+from enum import Enum
+
+import numpy as np
+
+from scaaml.aes_forward import AESSBOX
+from scaaml.stats.online import Mean, VarianceSinglePass
+
+
+class AttackPoint(Enum):
+    SUB_BYTES_IN = 0
+    SUB_BYTES_OUT = 0
+
+
+class LeakageModelAES128:
+    """The value that might be correlated with value(s) in the trace.
+    """
+
+    def __init__(self,
+                 byte_index: int = 0,
+                 attack_point: AttackPoint = AttackPoint.SUB_BYTES_IN,
+                 use_hamming_weight: bool = True) -> None:
+        """Gives the leakage function.
+
+        Args:
+
+          byte_index (int): Which byte to target (in range(16)).
+
+          attack_point (AttackPoint): Use either input or output of the first
+          SBOX.
+
+          use_hamming_weight (bool): Use just the Hamming weight of the value.
+        """
+        assert byte_index in range(16)
+        self._hamming_weight_table: list[int] = self._precompute_hw_table()
+        self._byte_index: int = byte_index
+        self._use_hamming_weight: bool = use_hamming_weight
+        self._attack_point: AttackPoint = attack_point
+
+    @staticmethod
+    def _precompute_hw_table() -> list[int]:
+        """Precompute Hamming weight table for values in range(256).
+        """
+        r = []
+        for i in range(256):
+            hwi = 0
+            x = i
+            while x > 0:
+                if x % 2 == 1:
+                    hwi += 1
+                x = x // 2
+            r.append(hwi)
+        return r
+
+    def leakage(self, plaintext: np.typing.NDArray[np.float64],
+                key: np.typing.NDArray[np.float64]) -> int:
+        """Return the leakage value.
+        """
+        # Get the byte value of the leakage.
+        byte_value: int
+        if self._attack_point == AttackPoint.SUB_BYTES_OUT:
+            byte_value = AESSBOX.sub_bytes_out(
+                key=bytearray(key),
+                plaintext=bytearray(plaintext),
+            )[self._byte_index]
+        elif self._attack_point == AttackPoint.SUB_BYTES_IN:
+            byte_value = AESSBOX.sub_bytes_in(
+                key=bytearray(key),
+                plaintext=bytearray(plaintext),
+            )[self._byte_index]
+        else:
+            raise NotImplementedError("Unknown attack point "
+                                      f"{self._attack_point}")
+
+        # Maybe convert to Hamming weight.
+        if self._use_hamming_weight:
+            return self._hamming_weight_table[byte_value]
+
+        return byte_value
+
+
+class SNRSinglePass:
+    """Single pass SNR implementation.
+    """
+
+    def __init__(self,
+                 leakage_model: LeakageModelAES128,
```

Comment:
Maybe abstract AES away?
```python
class LeakageModel(Protocol):
  def leakage(self, **params: Any) -> int:
    ...

```

And use that as a type instead of `LeakageModelAES128`.
This way the leakage models might be defined elsewhere than in the SNR as they might be useful for other computations

Reply:
When we have another leakage (e.g., ECC) definitely. Now it felt like over-engineering. Happy to add it if you think.

Reply:
Ack